include/dgl/heterograph_interface.h

```diff
@@ -0,0 +1,354 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file dgl/heterograph_interface.h
+ * \brief DGL heterogeneous graph index class.
+ */
+
+#ifndef DGL_HETEROGRAPH_INTERFACE_H_
+#define DGL_HETEROGRAPH_INTERFACE_H_
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <algorithm>
+
+#include "graph_interface.h"
+#include "array.h"
+
+namespace dgl {
+
+struct HeteroSubgraph;
+
+class HeteroGraphInterface;
+typedef std::shared_ptr<HeteroGraphInterface> HeteroGraphPtr;
+
+/*!
+ * \brief Heterogenous graph APIs
+ */
+class HeteroGraphInterface {
+ public:
+  /* \brief structure used to represent a list of edges */
+  // TODO(minjie): move this data structure outside of class definition so
+  // it can be shared by Graph and HeteroGraph.
+  typedef struct {
+    /* \brief the two endpoints and the id of the edge */
+    IdArray src, dst, id;
+  } EdgeArray;
+
+  virtual ~HeteroGraphInterface() = default;
+
+  ////////////////////////// query/operations on meta graph ////////////////////////
+  
+  /*! \return the number of vertex types */
+  virtual uint64_t NumVertexTypes() const = 0;
+
+  /*! \return the number of edge types */
+  virtual uint64_t NumEdgeTypes() const = 0;
+
+  /*! \return the meta graph */
+  virtual GraphPtr GetMetaGraph() const = 0;
```

Comment:
Does the C interface need meta data of the heterograph definition? if it is, how should meta graph be represented?